source/extensions/filters/common/ratelimit/ratelimit_registration.cc

```diff
@@ -0,0 +1,34 @@
+#include "extensions/filters/common/ratelimit/ratelimit_registration.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RateLimit {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(ratelimit_config);
+
+RateLimitServiceConfigPtr
+registerRateLimitServiceConfig(Server::Instance& server,
+                               const envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+  RateLimitServiceConfigPtr ratelimit_config =
+      server.singletonManager().getTyped<RateLimitServiceConfig>(
+          SINGLETON_MANAGER_REGISTERED_NAME(ratelimit_config), [&bootstrap] {
+            return std::make_shared<RateLimitServiceConfig>(bootstrap.rate_limit_service());
+          });
+  return ratelimit_config;
+}
+
+RateLimitServiceConfigPtr rateLimitConfig(Server::Configuration::FactoryContext& context) {
+  RateLimitServiceConfigPtr ratelimit_config =
+      context.singletonManager().getTyped<RateLimitServiceConfig>(
+          SINGLETON_MANAGER_REGISTERED_NAME(ratelimit_config), [] { return nullptr; });
```

Comment:
I think this can not be avoided. Even if we build client directly from core code, the filter config code would need a way to get the built client so there also same problem would come. I felt this is much cleaner and aligns with future code. I can add an ASSERT here to indicate that lambda should never be called.